Providers/Profile.swift

```diff
@@ -705,12 +715,28 @@ public class BrowserProfile: Profile {
 
             // We run this in the background after a few hundred milliseconds;
             // it doesn't really matter when it runs, so long as it doesn't
-            // happen in the middle of a sync. We take the lock to prevent that.
+            // happen in the middle of a sync.
+
+            let resetDatabase: () -> Success = { _ in
+                return self.handleRecreationOfDatabaseNamed(name) >>== { _ in
+                    log.debug("Reset of \(name) done")
+                    return succeed()
+                }
+            }
+
             self.doInBackgroundAfter(millis: 300) {
-                OSSpinLockLock(&self.syncLock)
-                self.handleRecreationOfDatabaseNamed(name).upon { res in
-                    log.debug("Reset of \(name) done: \(res.isSuccess)")
-                    OSSpinLockUnlock(&self.syncLock)
+                self.syncLock.lock()
+                defer { self.syncLock.unlock() }
+                // If we're syncing already, then wait for sync to end, 
+                // then reset the database on the same serial queue.
+                if let reducer = self.syncReducer where !reducer.isFilled {
+                    reducer.terminal.upon { _ in
+                        deferDispatchAsync(self.syncQueue, f: resetDatabase)
```

Comment:
I don't think you need `dDA` — you don't care about the return value, right? So just `dispatch_async` will suffice.